### PR TITLE
gpt function support

### PR DESCRIPTION
**Pull request by colombod**

### CODE REVIEW

File: src/Microsoft.DotNet.Interactive.AIUtilities.Tests/FunctionTests.cs [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+                                                  "a",
+                                                  "b",
+                                                  "c"
+                                                ]
+                                              }
+                                              """);
+    }
+
+    [Fact]
+    public void can_create_function_from_delegate()
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.AIUtilities.Tests/FunctionTests.cs:93]: This one seems less specific than the others and probably not needed.

File: src/Microsoft.DotNet.Interactive.AIUtilities.Tests/FunctionTests.cs [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        var function = GPTFunctionDefinition.Create((string a, double b) => $"{a} {b}", "concatString");
+
+        var jsonArgs = """
+                        {
+                            "name": "concatString",
+                            "arguments": "{ \"a\": \"Diego\", \"b\":123.0}"
+                        }
+                       """;
+        
+        function.Execute<string>(jsonArgs, out var result);
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.AIUtilities.Tests/FunctionTests.cs:230]: The return type can be inferred if the function definition type is generic.

File: src/Microsoft.DotNet.Interactive.AIUtilities/Function.cs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    internal GptFunction(Delegate function, string jsonSignature, string name)
+    {
+        Name = name;
+        _function = function;
+        JsonSignature = jsonSignature;
+    }
+
+    public object? Execute(string parameterJson)
+    {
+        // parameters extraction
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.AIUtilities/Function.cs:26]: Comment is unnecessary.

File: src/Microsoft.DotNet.Interactive.AIUtilities.Tests/FunctionTests.cs [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+                                                               "c"
+                                                             ]
+                                                           }
+                                                           """);
+    }
+
+    [Fact]
+    public void can_create_function_from_delegate_with_enums_as_parameters()
+    {
+        var declaration = GptFunction.Create((int a, double b, EnumType c) => $"{a} {b} {c}", "DoCompute");
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.AIUtilities.Tests/FunctionTests.cs:96]: I would either put the name property first or give it a default value. Why does the caller care about the function name?

Comment (colombod) [reply src/Microsoft.DotNet.Interactive.AIUtilities.Tests/FunctionTests.cs]: done

**Review by jonsequitur — APPROVED** [review]

**Pull request merged**